File: update.rb
require 'octokit'
require 'uri'
require 'json'
require 'open-uri'
require 'base64'

client = Octokit::Client.new(:access_token => ENV.fetch('GITHUB_TOKEN'))

data = JSON.parse(File.read("data.json"))

data["services"].each do |service|

  repo = service["github"].gsub("https://github.com/", "")

  package_path = service["packageLocation"].to_s + "package.json"

  begin
    package_json = JSON.parse(Base64.decode64(client.contents(repo, path: package_path).content))


    if package_json["dependencies"] && package_json["dependencies"]["nhsuk-frontend"]
      service["nhsukFrontendVersion"] = package_json["dependencies"]["nhsuk-frontend"]
    end

    if package_json["dependencies"] && package_json["dependencies"]["nhsuk-react-components"]
      service["nhsukReactComponentsVersion"] = package_json["dependencies"]["nhsuk-react-components"]
    end

  rescue Octokit::InvalidRepository
  rescue Octokit::NotFound
  rescue Octokit::SAMLProtected
  end

  lock_path = service["packageLocation"].to_s + "package-lock.json"

  begin
    package_lock_json = JSON.parse(Base64.decode64(client.contents(repo, path: lock_path).content))

    # npm v2/v3 lockfile format
    if package_lock_json["packages"] && package_lock_json["packages"]["node_modules/nhsuk-frontend"]
      service["nhsFrontendVersionPackageLock"] = package_lock_json["packages"]["node_modules/nhsuk-frontend"]["version"]
    # npm v1 lockfile format
    elsif package_lock_json["dependencies"] && package_lock_json["dependencies"]["nhsuk-frontend"]
      service["nhsFrontendVersionPackageLock"] = package_lock_json["dependencies"]["nhsuk-frontend"]["version"]
    end

  rescue Octokit::NotFound
    # Fall back to yarn.lock
    yarn_lock_path = service["packageLocation"].to_s + "yarn.lock"

    begin
      yarn_lock = Base64.decode64(client.contents(repo, path: yarn_lock_path).content)

      # Match a block starting with "nhsuk-frontend@..." and extract its resolved version
      if yarn_lock =~ /^"?nhsuk-frontend@[^:]+:?\n(?:.*\n)*?\s+version[:\s]+"?([^\s"]+)"?/
        service["nhsFrontendVersionPackageLock"] = $1
      end

    rescue Octokit::InvalidRepository
    rescue Octokit::NotFound
    rescue Octokit::SAMLProtected
    end

  rescue Octokit::InvalidRepository
  rescue Octokit::SAMLProtected
  end
end

# Update data file
File.open("data.json", 'w') do |file|
  file.write(JSON.pretty_generate(data))
  file.write "\n"
end


data["services"].sort! do |a, b|
  version_a = a["nhsukFrontendVersion"].to_s.gsub(/[\^\~]/, "")
  version_b = b["nhsukFrontendVersion"].to_s.gsub(/[\^\~]/, "")

  if version_a < version_b
    1
  elsif version_a > version_b
    -1
  else
    a["name"] <=> b["name"]
  end
end

# Update README.md
File.open("README.md", 'w') do |file|

  file.write "The following table shows the current version of [NHSUK Frontend](https://github.com/nhsuk/nhsuk-frontend) used by different services.\n\n"

  file.write "| Service | Frontend version | Installed version |\n"
  file.write "| :------ | -------------------: | -------------------: |\n"

  data["services"].each do |service|

    name = service["name"].to_s
    url = service["github"].to_s
    frontend_version = service["nhsukFrontendVersion"].to_s
    installed_version = service["nhsFrontendVersionPackageLock"].to_s

    file.write "| [#{name}](#{url}) | #{frontend_version} | #{installed_version} |\n"
  end

end

File: update.js
import { existsSync } from "node:fs";
import { readFile, writeFile } from "node:fs/promises";
import { Octokit } from "octokit";

// Load local dev credentials; in CI, GITHUB_TOKEN is already set and no .env file exists
if (existsSync(".env")) {
  process.loadEnvFile(".env");
}

const octokit = new Octokit({ auth: process.env.GITHUB_TOKEN });

async function getFileContent(owner, repo, path) {
  const response = await octokit.rest.repos.getContent({ owner, repo, path });
  return Buffer.from(response.data.content, "base64").toString("utf8");
}

const data = JSON.parse(await readFile("data.json", "utf8"));

const remainingServices = [];

for (const service of data.services) {
  const [owner, repo] = service.github.replace("https://github.com/", "").split("/");

  try {
    await octokit.rest.repos.get({ owner, repo });
  } catch (error) {
    // GitHub returns 404 (not 403) for both deleted repos and private ones the token can't see,
    // so that's the only reliable "gone" signal. 403 covers other things (rate limits, SAML/SSO
    // enforcement, org token policies) that don't mean the repo disappeared, so keep the entry.
    if (error.status === 404) {
      console.warn(`Removing "${service.name}": ${owner}/${repo} is no longer accessible (404)`);
      continue;
    }

    if (error.status === 403) {
      console.warn(`Skipping "${service.name}" this run: ${owner}/${repo} returned 403 (${error.message})`);
      remainingServices.push(service);
      continue;
    }

    throw error;
  }

  remainingServices.push(service);

  const packageLocation = service.packageLocation ?? "";

  try {
    const packageJson = JSON.parse(await getFileContent(owner, repo, `${packageLocation}package.json`));

    if (packageJson.dependencies?.["nhsuk-frontend"]) {
      service.nhsukFrontendVersion = packageJson.dependencies["nhsuk-frontend"];
    }

    if (packageJson.dependencies?.["nhsuk-react-components"]) {
      service.nhsukReactComponentsVersion = packageJson.dependencies["nhsuk-react-components"];
    }
  } catch (error) {
    // Ignore missing/inaccessible repos, same as the previous Ruby rescue blocks
    if (error.status !== 404 && error.status !== 403) throw error;
  }

  try {
    const packageLockJson = JSON.parse(await getFileContent(owner, repo, `${packageLocation}package-lock.json`));

    // npm v2/v3 lockfile format
    if (packageLockJson.packages?.["node_modules/nhsuk-frontend"]) {
      service.nhsFrontendVersionPackageLock = packageLockJson.packages["node_modules/nhsuk-frontend"].version;
    // npm v1 lockfile format
    } else if (packageLockJson.dependencies?.["nhsuk-frontend"]) {
      service.nhsFrontendVersionPackageLock = packageLockJson.dependencies["nhsuk-frontend"].version;
    }
  } catch (error) {
    if (error.status === 404) {
      // Fall back to yarn.lock
      try {
        const yarnLock = await getFileContent(owner, repo, `${packageLocation}yarn.lock`);

        // Match a block starting with "nhsuk-frontend@..." and extract its resolved version
        const match = yarnLock.match(/^"?nhsuk-frontend@[^:]+:?\n(?:.*\n)*?\s+version[:\s]+"?([^\s"]+)"?/m);

        if (match) {
          service.nhsFrontendVersionPackageLock = match[1];
        }
      } catch (yarnError) {
        if (yarnError.status !== 404 && yarnError.status !== 403) throw yarnError;
      }
    } else if (error.status !== 403) {
      throw error;
    }
  }
}

data.services = remainingServices;

// Update data file
await writeFile("data.json", JSON.stringify(data, null, 2) + "\n");

data.services.sort((a, b) => {
  const versionA = String(a.nhsukFrontendVersion ?? "").replace(/[\^~]/g, "");
  const versionB = String(b.nhsukFrontendVersion ?? "").replace(/[\^~]/g, "");

  if (versionA < versionB) return 1;
  if (versionA > versionB) return -1;
  return a.name.localeCompare(b.name);
});

// Update README.md
let readme = "The following table shows the current version of [NHSUK Frontend](https://github.com/nhsuk/nhsuk-frontend) used by different services.\n\n";
readme += "| Service | Frontend version | Installed version |\n";
readme += "| :------ | -------------------: | -------------------: |\n";

for (const service of data.services) {
  const name = service.name ?? "";
  const url = service.github ?? "";
  const frontendVersion = service.nhsukFrontendVersion ?? "";
  const installedVersion = service.nhsFrontendVersionPackageLock ?? "";

  readme += `| [${name}](${url}) | ${frontendVersion} | ${installedVersion} |\n`;
}

await writeFile("README.md", readme);
